File: frontend/src/services/outreach.ts
import api from '../api';
import { EmailAttachment, EmailLog } from './gmail';

export interface ExtractResult {
  recipients: string[];
  best?: string | null;
}

export interface ComposeResult {
  draft_id: string;
  subject: string;
  body: string;
  note: string;
  recipient?: string;
  recipients: string[];
  provider_used?: string;
}

export interface OutreachSendPayload {
  to: string;
  subject: string;
  body: string;
  cc?: string;
  bcc?: string;
  attachments?: EmailAttachment[];
  application_id?: string;
  company?: string;
  position?: string;
  job_description?: string;
  create_application?: boolean;
  override_guardrails?: boolean;
}

export interface OutreachSendResult {
  email_log: EmailLog;
  application_id?: string;
  warnings: string[];
}

export interface InboundReply {
  id: string;
  thread_id?: string;
  from_email?: string;
  subject?: string;
  category?: string;
  handled?: string;      // drafted | auto_replied | notified | sent | dismissed
  draft_reply?: string;
  reply_subject?: string;
  received_at?: string;
}

export interface InboundPollResult {
  handled: number;
  inbound_ready: boolean;
  enabled: boolean;
}

export const outreachService = {
  async extract(jobDescription: string): Promise<ExtractResult> {
    const res = await api.post<ExtractResult>('/outreach/extract', { job_description: jobDescription });
    return res.data;
  },

  async compose(payload: {
    job_description: string;
    recipient?: string;
    company?: string;
    preferred_provider?: string;
  }): Promise<ComposeResult> {
    // Composition is an LLM call — allow a generous timeout.
    const res = await api.post<ComposeResult>('/outreach/compose', payload, { timeout: 120000 });
    return res.data;
  },

  async send(payload: OutreachSendPayload): Promise<OutreachSendResult> {
    const res = await api.post<OutreachSendResult>('/outreach/send', payload);
    return res.data;
  },

  async listReplies(): Promise<InboundReply[]> {
    const res = await api.get<InboundReply[]>('/outreach/replies');
    return res.data;
  },

  async poll(): Promise<InboundPollResult> {
    const res = await api.post<InboundPollResult>('/outreach/poll');
    return res.data;
  },

  async sendReply(id: string, body?: string): Promise<EmailLog> {
    const res = await api.post<EmailLog>(`/outreach/replies/${id}/send`, { body });
    return res.data;
  },

  async dismissReply(id: string): Promise<void> {
    await api.post(`/outreach/replies/${id}/dismiss`);
  },
};

File: backend/src/models/outreach.py
from typing import List, Optional
from pydantic import BaseModel

from .gmail import EmailAttachment, EmailLogResponse


class ExtractRequest(BaseModel):
    job_description: str


class ExtractResponse(BaseModel):
    recipients: List[str] = []
    best: Optional[str] = None


class ComposeRequest(BaseModel):
    job_description: str
    recipient: Optional[str] = None
    company: Optional[str] = None
    preferred_provider: Optional[str] = None


class ComposeResponse(BaseModel):
    draft_id: str
    subject: str
    body: str
    note: str = ""
    recipient: Optional[str] = None
    recipients: List[str] = []
    provider_used: Optional[str] = None


class OutreachSendRequest(BaseModel):
    to: str
    subject: str
    body: str
    cc: Optional[str] = None
    bcc: Optional[str] = None
    attachments: List[EmailAttachment] = []
    application_id: Optional[str] = None
    company: Optional[str] = None
    position: Optional[str] = None
    job_description: Optional[str] = None
    create_application: bool = True
    override_guardrails: bool = False  # resend past the already-emailed / daily-cap guard


class OutreachSendResponse(BaseModel):
    email_log: EmailLogResponse
    application_id: Optional[str] = None
    warnings: List[str] = []


class InboundReplyResponse(BaseModel):
    id: str
    thread_id: Optional[str] = None
    from_email: Optional[str] = None
    subject: Optional[str] = None
    category: Optional[str] = None
    handled: Optional[str] = None       # drafted | auto_replied | notified | sent | dismissed
    draft_reply: Optional[str] = None
    reply_subject: Optional[str] = None
    received_at: Optional[str] = None


class SendReplyRequest(BaseModel):
    body: Optional[str] = None          # optional edited body; falls back to the stored draft


class InboundPollResponse(BaseModel):
    handled: int = 0
    inbound_ready: bool = True
    enabled: bool = True
